database/src/main/resources/db/migration/V1_24__Add_gas_fee_stats.sql

```diff
@@ -0,0 +1,191 @@
+     )
+INSERT
+INTO tx_single_message_cache (tx_timestamp, tx_hash, gas_used, tx_message_type)
+SELECT *
+FROM STATS
+ON CONFLICT DO NOTHING;
+
+-- Daily aggregate of historical data
+INSERT
+INTO tx_single_message_gas_stats_day
+SELECT date_trunc('DAY', tx_timestamp)           AS tx_timestamp,
+       min(gas_used)                             AS min_gas_used,
+       max(gas_used)                             AS max_gas_used,
+       round(avg(gas_used))                      AS avg_gas_used,
+       coalesce(round(stddev_samp(gas_used)), 0) AS stddev_gas_used,
+       tx_message_type
+FROM tx_single_message_cache
+GROUP BY date_trunc('DAY', tx_timestamp),
+         tx_message_type
+ON CONFLICT DO NOTHING;
+
+-- Hourly aggregate of historical data
+INSERT
+INTO tx_single_message_gas_stats_hour
+SELECT date_trunc('HOUR', tx_timestamp)          AS tx_timestamp,
+       min(gas_used)                             AS min_gas_used,
+       max(gas_used)                             AS max_gas_used,
+       round(avg(gas_used))                      AS avg_gas_used,
+       coalesce(round(stddev_samp(gas_used)), 0) AS stddev_gas_used,
+       tx_message_type
+FROM tx_single_message_cache
+GROUP BY date_trunc('HOUR', tx_timestamp),
+         tx_message_type
+ON CONFLICT DO NOTHING;
+
+-- Update daily gas stats with latest data
+CREATE OR REPLACE PROCEDURE update_daily_gas_stats()
+    LANGUAGE plpgsql
+AS
+$$
+DECLARE
+    unprocessed INT[];
+BEGIN
+    -- Collection unprocessed record IDs
+    SELECT array_agg(id) FROM tx_single_message_cache WHERE processed = false INTO unprocessed;
+
+    -- Upsert unprocessed stats.
+    INSERT
+    INTO tx_single_message_gas_stats_day(tx_timestamp,
+                                         min_gas_used,
+                                         max_gas_used,
+                                         avg_gas_used,
+                                         stddev_gas_used,
+                                         tx_message_type)
+    SELECT date_trunc('DAY', tx_timestamp)          AS tx_timestamp,
+           min(gas_used)                             AS min_gas_used,
+           max(gas_used)                             AS max_gas_used,
+           round(avg(gas_used))                      AS avg_gas_used,
+           coalesce(round(stddev_samp(gas_used)), 0) AS stddev_gas_used,
+           tx_message_type
+    FROM tx_single_message_cache
+    WHERE id = ANY (unprocessed)
+    GROUP BY date_trunc('DAY', tx_timestamp),
+             tx_message_type
+    ON CONFLICT (tx_timestamp, tx_message_type)
+        DO UPDATE
+        SET min_gas_used    = tx_single_message_gas_stats_day.min_gas_used + excluded.min_gas_used,
+            max_gas_used    = tx_single_message_gas_stats_day.max_gas_used + excluded.max_gas_used,
+            avg_gas_used    = tx_single_message_gas_stats_day.avg_gas_used + excluded.avg_gas_used,
+            stddev_gas_used = tx_single_message_gas_stats_day.stddev_gas_used + excluded.stddev_gas_used;
```

Comment:
This seems wrong for the values we are trying to update. min should be min, max should be max, etc.  Maybe ON CONFLICT should reprocess the given hour?

Reply:
Resolved offiline